File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] costs) {
        int answer = 0;
        boolean[] visited = new boolean[n];
        Map<Integer, List<int[]>> map = new HashMap<>();
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        
        for(int[] cost : costs){
            int start = cost[0], end = cost[1];
            map.computeIfAbsent(start, k -> new ArrayList<>()).add(new int[]{end, cost[2]});
            map.computeIfAbsent(end, k-> new ArrayList<>()).add(new int[]{start, cost[2]});
        }
        pq.offer(new int[]{0,0});
        
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            int node = cur[0], cost = cur[1];
            if(visited[node]) continue;
            visited[node] = true;
            answer+=cost;
            for(int[] neighbor : map.get(node)){
                int n_node = neighbor[0], n_cost = neighbor[1];
                if(!visited[n_node]) pq.offer(new int[]{n_node, n_cost});
            }
        }
        return answer;
    }
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
    vector visited(n, false);
    unordered_map<int, vector<pair<int,int>>> map;
    priority_queue<pair<int,int>, vector<pair<int, int>>, greater<>> pq;

    for (const auto& cost : costs)
    {
        int from = cost[0], to = cost[1], c = cost[2];
        map[from].emplace_back(c,to);
        map[to].emplace_back(make_pair(c, from));
    }

    pq.emplace(0,0);
    while (!pq.empty())
    {
        pair<int, int> cur = pq.top(); pq.pop();
        int cost = cur.first, node = cur.second;
        if (visited[node]) continue;
        visited[node] = true;
        answer += cost;
        for (const auto& next : map[node])
        {
            int next_cost = next.first, next_node = next.second;
            if (!visited[next_node]) pq.emplace(next_cost, next_node);
        }

    }

    return answer;
}